mijia/examples/characteristics.rs

```diff
@@ -1,37 +1,78 @@
-use mijia::bluetooth::BluetoothSession;
-use std::time::Duration;
-use tokio::time;
-
-const SCAN_DURATION: Duration = Duration::from_secs(5);
+use mijia::bluetooth::{BleUuid, BluetoothSession};
+use std::ops::RangeInclusive;
+use std::str;
 
 #[tokio::main]
 async fn main() -> Result<(), eyre::Report> {
     pretty_env_logger::init();
 
     let (_, session) = BluetoothSession::new().await?;
 
-    // Start scanning for Bluetooth devices, and wait a while for some to be discovered.
-    session.start_discovery().await?;
-    time::sleep(SCAN_DURATION).await;
-
     // Get the list of devices whose services are currently known and print them with their
     // characteristics.
     let devices = session.get_devices().await?;
     println!("Devices:");
     for device in devices {
         let services = session.get_services(&device.id).await?;
         if !services.is_empty() {
-            println!("{}: {:?}", device.mac_address, device.id);
-            println!("Services: {:#?}", services);
+            println!("{}: {}", device.mac_address, device.id);
             for service in services {
-                let characteristics = session.get_characteristics(&service.id).await?;
                 println!(
-                    "Service {:?} characteristics: {:#?}",
-                    service, characteristics
+                    "Service {} ({}): {}",
+                    service.uuid.to_short_string(),
+                    if service.primary {
+                        "primary"
+                    } else {
+                        "secondary"
+                    },
+                    service.id
                 );
+                let characteristics = session.get_characteristics(&service.id).await?;
+                for characteristic in characteristics {
+                    println!(
+                        "  Characteristic {}: {}",
+                        characteristic.uuid.to_short_string(),
+                        characteristic.id
+                    );
+                    let descriptors = session.get_descriptors(&characteristic.id).await?;
+                    for descriptor in descriptors {
+                        println!(
+                            "    Descriptor {}: {}",
+                            descriptor.uuid.to_short_string(),
+                            descriptor.id
+                        );
+                        let value = session.read_descriptor_value(&descriptor.id).await?;
+                        println!("      {}", debug_format_maybe_string(&value));
+                    }
+                }
             }
         }
     }
 
     Ok(())
 }
+
+const PRINTABLE_ASCII_RANGE: RangeInclusive<u8> = 0x20..=0x7E;
```

Comment:
does this work?

```suggestion
const PRINTABLE_ASCII_RANGE: RangeInclusive<u8> = ' '..='~';
```

Reply:
No, because those literals are `char`s not `u8`s.